src/dos/cdrom_win32.cpp

```diff
@@ -356,34 +359,92 @@ bool CDROM_Interface_Win32::HasDataTrack() const
+	// was implemented
+
+	constexpr uint32_t MinAllowedFramesPerCall = 1;
+	// Chosen to limit the number of failed IOCTL requests for known drives
+	constexpr uint32_t MaxAllowedFramesPerCall = 130;
+
+	const auto target_frames = std::min(frames_requested, MaxAllowedFramesPerCall);
+	while (true)
+	{
+		uint32_t num_frames = target_frames;
+		if (frames_per_call_known_bad != 0 &&
+	            num_frames >= frames_per_call_known_bad) {
+	            	// The number of frames we are going to request is
+	            	// definitely too much for the particular drive;
+	            	// get closer to the known good value (binary search)
+			num_frames = (frames_per_call_known_bad + frames_per_call_known_good) / 2;
+			num_frames = std::max(num_frames, MinAllowedFramesPerCall);
+			num_frames = std::min(num_frames, target_frames);
+		}
+
+		audio_frames.resize(num_frames * SAMPLES_PER_REDBOOK_FRAME);
+
+		RAW_READ_INFO read_info = {};
+
+		read_info.DiskOffset.LowPart = sector * BYTES_PER_COOKED_REDBOOK_FRAME;
+		read_info.SectorCount        = num_frames;
+		read_info.TrackMode          = CDDA;
+
+		DWORD control_code       = IOCTL_CDROM_RAW_READ;
+		LPVOID input_buffer      = &read_info;
+		DWORD input_buffer_size  = sizeof(read_info);
+		LPVOID output_buffer     = audio_frames.data();
+		DWORD output_buffer_size = audio_frames.size() * sizeof(int16_t);
+		LPDWORD bytes_returned   = NULL;
+		LPOVERLAPPED overlapped  = NULL;
+
+		const bool result = DeviceIoControl(cdrom_handle,
+		                                    control_code,
+		                                    input_buffer,
+		                                    input_buffer_size,
+		                                    output_buffer,
+		                                    output_buffer_size,
+		                                    bytes_returned,
+		                                    overlapped);
+
+		if (result) {
+			// Success, audio frames received
+			frames_per_call_known_good = std::max(num_frames,
+				frames_per_call_known_good);
+			return audio_frames;
+		}
+
+		if (GetLastError() == ERROR_INVALID_PARAMETER &&
+		    num_frames > MinAllowedFramesPerCall) {
+			// We have requested too many frames for the driver
+			// to handle within a single call; retry, never again
+			// request that many
+			frames_per_call_known_bad = num_frames;
+
+			// The algorithm here is written with the assumption
+			// that the allowed number of frames remains stable for
+			// the device
+			if (frames_per_call_known_bad != 0 &&
+			    frames_per_call_known_bad <= frames_per_call_known_good) {
+			    	frames_per_call_known_good = 0;
+
+			    	// This shouldn't happen, attract attention here
+			    	static bool already_warned = false;
+			    	if (!already_warned) {
+					LOG_ERR("CDROM_IOCTL: Valid number of frames changed");
+					already_warned = true;
+				}
+			}
+		} else {
+			// Some other error, audio frames should contain silence
+			return audio_frames;
+		}
+	}
```

Comment:
Are there any drawbacks to setting this to a know-good low-enough value, or 27, or day even going lower to 20?

I guess I prefer the simplest solution that does the job 😄

Reply:
This was one of my proposals, but...

Problem 1: We have no idea what the safe minimum is, we only have test results from my system, no clear documentation, no insight how the driver stack is organized, no OS source code access. I can fully imagine the maximum being even lower on some more exotic USB controllers, ARM laptops, etc.

Problem 2: I have almost 0 experience with the desktop Windows programming, but on Windows CE I have seen WinAPI calls sporadically taking a lot of time to return, without any obvious reasons. So I would prefer to limit the amount of calls if we can.

@weirddan455 What do you think? If you think it's safe, I'm perfectly fine with dropping this solution and just using lower frame limit (maybe with some LOG_ERR to immediately draw attention to the proper place if the IOCTL fails).

Reply:
One thought is that we should try to avoid large and infrequent "bursty" reads and instead try for more frequent and smaller "smooth" reads, because the former will cause fast drives to spin up.

Listening to music with a drive running at 1x is a lot more pleasant than a drive running at 16x or faster.

The original CDROM drives only spun at a cool and quiet 1x (~150 KB/s). To keep up with realtime, we only have to read 75 CDDA frames per second or about one every every 13 ms.

With @weirddan455's threading and queuing, maybe we should try for just 1 frame per read as this is probably the closest read profile to encourage the drive to relax down to it's lowest rotation speed. (this is just a hunch.. unfortunately I don't have hardware to try it out).

Reply:
@interloper98 This is indeed a valid point to consider.

I have tested both my USB drives, but I could hear no noise pattern change when reading just a single frame. With both devices the drive spins up and slows down repeatedly. Probably the only think I have changed was to empty the internal OS buffers in multiple calls vs in one go.

As for the internal drive, I can't tell, I have a rather power hungry CPU (which means the fans make quite some noise), and the drive is not one of the dirt-cheap ones, so the disc doesn't vibrate.

Reply:
> With both devices the drive spins up and slows down repeatedly. Probably the only think I have changed was to empty the internal OS buffers in multiple calls vs in one go.

Nice tests @FeralChild64 ! Yes, that yo-yo'ing is painful but it sounds like number of frames isn't going to get that under control on it's own. 

I had one of those 'Mad Dog 70x' drives in the early 2000s and it was my mission to find applications to lock that thing to it's slowest rate of something like 8x or 12x.. because yo-yo'ing from 70x to idle on every game read was madness! I guess that's why it was called the 'Mad Dog'.  :man_facepalming: 

> Manipulating the CD-ROM speed has to be done using special IOCTL’s - on Windows these are IOCTL_CDROM_GET_PERFORMANCE / IOCTL_CDROM_SET_SPEED, on Linux these are CDROM_SELECT_SPEED

Excellent. Some logic could be: set the drive speed to the slowest if it's an audio-only CD (no data track), otherwise (data or mixed-mode) set the speed to something reasonable for the 90s DOS era, like 8x or 12x.

Reply:
8x would be good since its still CAV:

![drive](https://github.com/user-attachments/assets/3c3dde47-2a73-4b97-9dcc-a4cb1778a0e3)

Reply:
https://github.com/dosbox-staging/dosbox-staging/issues/3842 - now let's back to the main discussion here.

Reply:
Based on your results comparing multi-frame vs single frame, my vote would be for a frame size of 1 given we're dealing with audio (which is the lowest possible CD rate) and given the thread + queue is keeping our outbound buffer full.

Even if we're draining an intermediate buffer (driver or OS), fetching one at a time gives the buffer a change to refill from the hard at various levels (say, 50% full).

Reply:
If we can find a "known safe value" that works for all drives, that would be the simplest solution.  I've come up empty so far trying to research that though.

@FeralChild64 's solution seems reasonable, though IMO it is a bit noisy having all this binary search logic inside the `ReadAudio()` function.

Possible alternative:  Can we do a test to determine max frames a drive can read inside `Open()` by reading the first audio track?  I have a hunch that the Win32 call will fail immediately if the `SectorCount` value is out of range.  If that's the case, we can do a linear test starting at some maximum (55 or higher if you think some drives can handle it) and decrementing by 1 if that doesn't work.  Take the first value that succeeds and store that as the maximum to use in `ReadAudio()`.

This would simplify the code quite a bit and we only have to test once when the drive initially gets mounted.  Assuming the test doesn't take too long to execute, I think it would be viable.

Reply:
We would need to read the TOC and determine a suitable audio track for testing. Maybe it would be easier to do this within the first ReadAudio call?
At this point we at least know where the first audio track starts.

Reply:
Anything in the sub-second range of buffering is reasonable (4 KB to ~128 KB ), because the hardware can hand that back to us with just a couple revolutions of the media, and these small chunks are unlikely to cause big latency stalls on the drive.

(So I'll modify my vote for anything that gives us those :-)

(The flip side would be searching for the maximum buffer with a cap: _"Hey, my BD drive support 256 MB reads!"_, which is the high-latency dilemma the network industry got itself into with runaway [bufferbloat](https://en.wikipedia.org/wiki/Bufferbloat). The concern is when events in the game change and you need to seeking to another chunk of audio or switching tracks - refilling several MB of buffer hurts latency).

Reply:
>  We would need to read the TOC and determine a suitable audio track for testing. Maybe it would be easier to do this within the first ReadAudio call?

I already am reading the TOC inside `Open()` as kind of an arbitrary ioctl to test whether the requested drive is actually a CDROM drive.  It would probably be fine to test in the first `ReadAudio()` call as well but we don't have TOC information stored there either.  Just a sector number which could theoretically be at the end of the disk with a read request less than our maximum.

Reply:
OK, indeed. Is there a reason why isn’t the TOC cached?

Reply:
>  OK, indeed. Is there a reason why isn’t the TOC cached?

Simplicity I guess.  We don't need the TOC for anything internal.  We just read it on demand when the DOS application requests track information (which is fairly infrequently I believe).  For the frequent `ReadAudio()` calls, all we need is a sector number and a length.

Reply:
But the routine to convert the values from TOC to start track sector is somewhere else, right? The `CDROM_Interface_Win32` class does not seem to have this knowledge (which is good, since this code should be shared across the platforms).

I would prefer not to do any refactoring here.

Reply:
>  But the routine to convert the values from TOC to start track sector is somewhere else, right?

All of that information is in the TOC.  See `GetAudioTracks()` and `GetAudioTrackInfo()` for how it is extracted.  The Win32 call does return track information in Minute/Second/Frame format rather than sector.  There's some code in the MSCDEX layer to convert that though (on the DOS side, programs can request in either format).

https://github.com/dosbox-staging/dosbox-staging/blob/489f43010fba7c8ac1d11a1f7981c9da99fe989e/src/dos/dos_mscdex.cpp#L490-L500

There's a chance you can get away without fussing with that though.  All we care about for the test is what `SectorCount` values the CDROM driver will accept.  If the `IOCTL_CDROM_RAW_READ` call will succeed reading from a data track, we can just test reading from the first sector on the disc.

Reply:
Ok, so how about we aim for simplicity and hardcode the value to 1 and see how that goes? It cannot get safer than that and reading at redbook audio rates is rather undemanding. I can also try to test it with USB drives, I think I have two of them somewhere. Then we reconsider this if people report issues with CD Audio .

Yeah I always used a CD speed throttler utility in Windows because I hated the hoover sound when playing games on 32x-40x drives 😆

Reply:
Exactly. @weirddan455, what you are proposing, far exceeds the effort I am willing to spend here. I’m not digging into MSCDEX (I have already too many complex or time consuming topics started), I am not doing any non-trivial refactoring here, and I am definitely not willing to take the risk to read some random sectors using IOCTL which is intended by Microsoft to read the CDDA tracks only.

I can either do something similar to this PR, or just reduce the max number of sectors read. Anything more = someone else’s job :)

Reply:
It was just a suggestion.  You don't need to dig into MSCDEX or refactor anything.  The conversion from MSF to Sector was all I was pointing out with that MSCDEX function.  I could write a simple `TestMaxSector()` function and it wouldn't be that much code.  I don't mind doing it if you want to hand this off to me.

>  Ok, so how about we aim for simplicity and hardcode the value to 1 and see how that goes?

Or just do this.  I'm slightly concerned if slower drives can keep up.  I did some timing on this back in my original single-threaded implementation.  A value of 1 is a lot more inefficient than, say 20.  More total time is spent inside the IOCTL calls (presumably due to overhead of making 20x more system calls).  Going from 55 down to 20 is a lot less noticeable.

>  I agree with @interloper98 on bursty reads being potentially problematic.

Bursty reads were problematic in the single-threaded implementation.  They would stall the main thread for too long so I had to lower the sector count from the maximum.  It's not really a problem with the way this threaded reader is implemented.  All we really care about is throughput which is why I raised the count back to maximum.  If the drive can keep up reading audio in real-time, everyone's happy.

Reply:
tl;dr on my last comment:  I'm fine with lowering the value to 1 and re-evaluating later if someone complains.

Reply:
Threaded queue, low-latency backlog, and drive RPM reduction (in @FeralChild64's other ticket) is going to be the deluxe way to play DOS CD games :rocket:

Reply:
Latency isn't really a concern with CDROM audio in general.  The `RWQueue` uses a large 5 second buffer just because I can and it helps with intermittent spikes in read time.  That gets filled with a bursty read when a track starts playing (and it still will with reading a single frame at a time, it'll just take more syscalls to do it).

The mixer doesn't wait until the buffer is fully filled to start playing though.  It'll start as soon as the queue has a minimum number of frames (usually 1-10ms worth of audio).  On track change, the queue gets immediately cleared so it doesn't wait to fully drain either.

Reply:
> Threaded queue, low-latency backlog, and drive RPM reduction (in @FeralChild64's other ticket) is going to be the deluxe way to play DOS CD games 💿 🚀

Yeah, lot of effort for something that's well on the way out (physical CD-ROM drives), but why not if it doesn't complicate the code too much 😆

Reply:
Optical drives. _My precious…_

Reply:
Reduced the value to a safe one, dropped the guessing code.